### PR TITLE
NPU Mean Op

### PR DESCRIPTION
**Pull request by OleNet**
### PR types

New features
### PR changes

OPs
### Describe

[NPU] add op for mean 

image

image

### CODE REVIEW

**Comment by OleNet** [comment]
image
image

**Review by liym27 — CHANGES_REQUESTED** [review]
1. 将单测验证截图，移到 PR 描述中。
2. 补充 npu kernel 被调用的截图，否则无法确定 npu kernel 是否被调用。

File: paddle/fluid/operators/mean_op_npu.cc [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                  {"keep_dims", false},
+                  {"axes", axes}};
+
+    std::vector<int64_t> out_dims;
+    out_dims.push_back(1);
+    out->Resize(framework::make_ddim(out_dims));
+    out->mutable_data<T>(ctx.GetPlace());
+
+    Tensor reduced_out(x->type());
+    std::vector<int64_t> reduced_dout_dims;
```
Comment (pangyoki) [inline paddle/fluid/operators/mean_op_npu.cc:43]: `reduced_dout_dims`这个变量有什么用吗？

File: paddle/fluid/operators/mean_op_npu.cc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      axes.push_back(i);
+    }
+
+    framework::NPUAttributeMap attr_input = {
+                  {"keep_dims", false},
+                  {"axes", axes}};
+
+    std::vector<int64_t> out_dims;
+    out_dims.push_back(1);
+    out->Resize(framework::make_ddim(out_dims));
```
Comment (pangyoki) [inline paddle/fluid/operators/mean_op_npu.cc:39]: InferShape阶段，输出的shape已经设置为{1}了，不需要重复设置

Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc:39]: Don't need to resize tensor `out`, which has been set size in `InferShape` .

**Review by zhiqiu — APPROVED** [review]
LGTM

File: paddle/fluid/operators/mean_op_npu.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    out->mutable_data<T>(ctx.GetPlace());
+
+    Tensor reduced_out(x->type());
+    std::vector<int64_t> reduced_dout_dims;
+    reduced_dout_dims.push_back(1);
+    reduced_out.Resize(framework::make_ddim(reduced_dout_dims));
+    reduced_out.mutable_data<T>(ctx.GetPlace());
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc:46]: It seems that this tensor `reduced_out` has never been used.

File: paddle/fluid/operators/mean_op_npu.cc [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    auto reduce_ndim = x->dims().size();
+    std::vector<int> axes;
+    for (auto i = 0; i < reduce_ndim; ++i) {
+      axes.push_back(i);
+    }
+
+    framework::NPUAttributeMap attr_input = {
+                  {"keep_dims", false},
+                  {"axes", axes}};
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc:35]: According to `MeanParser`, `axes` should be  an empty vector.

File: paddle/fluid/operators/mean_op_npu.cc [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    IG->mutable_data<T>(context.GetPlace());
+
+    // ones
+    Tensor ones(grad->type());
+    std::vector<int64_t> dout_dims;
+    for (auto i = 0; i < IG->dims().size(); ++i) {
+      dout_dims.push_back(IG->dims()[i]);
+    }
+    ones.Resize(framework::make_ddim(dout_dims));
+    ones.mutable_data<T>(context.GetPlace());
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc:89]: lines 84 - 89, can be replaced by one line:
```
ones.mutable_data<T>(IG->dims(), context.GetPlace());
```

File: paddle/fluid/operators/mean_op_npu.cc [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    // means mul ones
+    Tensor mean_ma(grad->type());
+    mean_ma.Resize(framework::make_ddim(dout_dims));
+    mean_ma.mutable_data<T>(context.GetPlace());
+    auto runner_mul_1 = NpuOpRunner("Mul", {mean_tensor, ones}, {mean_ma}, {});
+    runner_mul_1.Run(stream);
+
+    // and mul grad
+    auto runner_mul_2 = NpuOpRunner("Mul", {mean_ma, *grad}, {*IG}, {});
+    runner_mul_2.Run(stream);
```
Comment (liym27) [inline paddle/fluid/operators/mean_op_npu.cc:112]: Lines 103-112, why not use only one "Mul", which input are "mean_tensor" and "*grad", output is "*IG"?

**Review by liym27 — COMMENTED** [review]
For `MeanGradNPUKernel`, please better to call ascend ops according to `MeanGradParser`  https://github.com/PaddlePaddle/Paddle/blob/ascendrc/python/paddle/distributed/fleet/meta_optimizers/ascend/ascend_parser.py#L2048

**Pull request merged**